src/test/rgw/rgw_multi/tests.py

```diff
@@ -2859,3 +2884,76 @@ def test_sync_single_bucket_to_multiple():
     remove_sync_policy_group(c1, "sync-bucket", bucketA.name)
     remove_sync_policy_group(c1, "sync-group")
     return
+
+def test_topic_notification_sync():
```

Comment:
please add a test that checks the actual notifications. 
e.g. use an http endpoint verify that the number of notifications we see there is number of zones X number of object put ops

Reply:
I was thinking of adding that test inside test_bn.py, this test suite is more about testing the replications.
regarding adding the test test_bn.py is there a way to restart the gateway from within the test. (to enable the feature), in multisite world running `period commit` turns on the feature flag but in single site world we have to restart the rgw, so if we have a way to restart the gateway checking the notification count inside the test_bn.py would make more sense

Reply:
> I was thinking of adding that test inside test_bn.py, this test suite is more about testing the replications. regarding adding the test test_bn.py is there a way to restart the gateway from within the test. (to enable the feature),

nothing that I know of, but I can look into that. adding this would be useful for other notification tests as well that are currently manual.

> in multisite world running `period commit` turns on the feature flag but in single site world we have to restart the rgw, so if we have a way to restart the gateway checking the notification count inside the test_bn.py would make more sense

what about real single zone deployment. will we have to restart all RGWs when we change this flag?

Reply:
> > I was thinking of adding that test inside test_bn.py, this test suite is more about testing the replications. regarding adding the test test_bn.py is there a way to restart the gateway from within the test. (to enable the feature),
> 
> nothing that I know of, but I can look into that. adding this would be useful for other notification tests as well that are currently manual.
> 
> > in multisite world running `period commit` turns on the feature flag but in single site world we have to restart the rgw, so if we have a way to restart the gateway checking the notification count inside the test_bn.py would make more sense
> 
> what about real single zone deployment. will we have to restart all RGWs when we change this flag?

currently `stop.sh` does  has a way to stop the gateway, it by defaults stops all gateways but there is no provision to start the rgw as such inside vstart. For multisite there is a file `mrgw.sh` that does start the rgw but for mstart clusters, may that could be made more generic to work with single site as well and then we could have a way to stop and start rgw.

@cbodley regarding the comment `what about real single zone deployment. will we have to restart all RGWs when we change this flag?` i am not sure, but i think if one of the gateway is restarted it will pick up the new flag

Reply:
> i am not sure, but i think if one of the gateway is restarted it will pick up the new flag

you can have a single zone deployment with many RGWs. will you have to restart all of them to change the parameter?
how will you do that from as a  `radosgw-admin` command?

Reply:
> > i am not sure, but i think if one of the gateway is restarted it will pick up the new flag
> 
> you can have a single zone deployment with many RGWs. will you have to restart all of them to change the parameter? how will you do that from as a `radosgw-admin` command?

Not via admin command, there is a script mrgw.sh that needs to point to correct ceph.conf and that can be used for start and vstart.sh for stopping. Only thing that we need inside test is to know the count of rgw instances n their port no.

Reply:
my concern is not about the test. it is about a real single-zone deployment (with multiple RGWs).
i nthe migration process we will have to change this parameter, and this should be done as part of the `radosgw-admin` command that does the migration.
will we have to reload all RGWs in this case?

Reply:
> my concern is not about the test. it is about a real single-zone deployment (with multiple RGWs). i nthe migration process we will have to change this parameter, and this should be done as part of the `radosgw-admin` command that does the migration. will we have to reload all RGWs in this case?

yep.. thats why earlier i had tagged  @cbodley as well, regarding the behavior in single site with multiple rgw instances. My understanding was we have to reload all the gateway and thats what i tested in my testing. I did not check what if some are restarted and others are not, what will be the behavior ?
but to your question, after doing the `radosgw-admin zone modify` n `radosgw-admin zonegroup modify` we have to restart all the gateways.

Reply:
> @cbodley regarding the comment `what about real single zone deployment. will we have to restart all RGWs when we change this flag?` i am not sure, but i think if one of the gateway is restarted it will pick up the new flag

if the single-site deployment has a realm, then you use `period commit` and rgws will automatically do 'realm reload' to pick up the new configuration. if there isn't a realm, then each radosgw will have to restart to load the new zone/zonegroup configuration

i'm pretty sure that cephadm does always create a realm for single-site deployments, but we generally won't have one in vstart or teuthology deployments

Reply:
> > @cbodley regarding the comment `what about real single zone deployment. will we have to restart all RGWs when we change this flag?` i am not sure, but i think if one of the gateway is restarted it will pick up the new flag
> 
> if the single-site deployment has a realm, then you use `period commit` and rgws will automatically do 'realm reload' to pick up the new configuration. if there isn't a realm, then each radosgw will have to restart to load the new zone/zonegroup configuration
> 
> i'm pretty sure that cephadm does always create a realm for single-site deployments, but we generally won't have one in vstart or teuthology deployments

correct, vstart does not have realm, as when i try to do `period commit` it complains no realm present.
So for actual world users of single zone, there will always be realm and `migration` can call the `period commit` once all the rgw's have migrated to new version ?

Reply:
maybe i can add that to the bucket notification test suite?

Reply:
> maybe i can add that to the bucket notification test suite?

you mean adding a realm while creating the vstart or creating the ream from within the test ?
i could do that as part of this PR?

Reply:
> > maybe i can add that to the bucket notification test suite?
> 
> you mean adding a realm while creating the vstart or creating the realm from within the test ? i could do that as part of this PR?

i mean, add that to the bucket notifications test suite. so we have that both with local vstart runs and with teuthology

Reply:
so, the test plan is to (both locally and in teuthology):
* be able to run the **notification test suite** against a v1 multisite setup (with single zone), both locally and remotely - this is just a preliminary step, and could be used for most basic regression of this PR
* run the **notification test suite** against a v2 multisite (with single zone) setup - this is main regression for this PR
* add a test to the **multisite test suite** that verifies that notification are sent from a 2ndary zone after topic and notification are configured on the master zone (and vice versa)